### PR TITLE
test(robot): extend negative test cases to include v2 data engine

### PR DESCRIPTION
**Pull request by yangchiu**
#### Which issue(s) this PR fixes:

Issue N/A

#### What this PR does / why we need it:

extend negative test cases to include v2 data engine

### CODE REVIEW

**Comment by yangchiu** [comment]
@roger-ryao @chriscchien Please help merge this to enhance the test coverage for v2 volume. Thank you!

**Comment by yangchiu** [comment]
https://github.com/longhorn/longhorn/issues/9171#issuecomment-2267541293

> The is a known issue. v2 volume doesn't support auto-salvage. The feature will be implemented in v1.8.0.

v2 volume is not possible to pass negative tests. Defer this PR for now.

**Review by khushboo-rancher — CHANGES_REQUESTED** [review]
nit: In general, if we can add brief description of test cases that will be helpful as the test cases are becoming bigger. 
e.g.
```
This test validates the cluster's stability and recovery during heavy data writing operations across multiple workloads with different storage configurations. The test creates:

3 RWO volumes with the following storage classes:
 a. longhorn-test
 b. strict-local
 c. strict-local (for statefulset)
3 RWX volumes with the following storage classes:
 a. longhorn-test
 b. nfs-4-2
 c. nfs-4-2 (for statefulset)
 d. nfs-hard-mount
 e. nfs-soft-mount
Deployments and statefulsets are created using these volumes, and data is continuously written to them. The cluster is then restarted, and the test verifies that all workloads resume correctly and data integrity is maintained post-restart.
```

File: e2e/libs/volume/crd.py [hunk 1]
```diff
@@ -418,7 +418,9 @@ def check_data_checksum(self, volume_name, data_id):
         expected_checksum = self.get_data_checksum(volume_name, data_id)
         actual_checksum = self.get_checksum(volume_name)
         logging(f"Checked volume {volume_name} data {data_id}. Expected checksum = {expected_checksum}. Actual checksum = {actual_checksum}")
-        assert actual_checksum == expected_checksum
+        if actual_checksum != expected_checksum:
+            logging(f"Checked volume {volume_name} data {data_id} failed. Expected checksum = {expected_checksum}. Actual checksum = {actual_checksum}")
+            time.sleep(self.retry_count)
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/libs/volume/crd.py:423]: Failing the test case doesn't help investigate and debug the issue. It only causes developers to spend more time reproducing it.

File: e2e/keywords/network.resource [hunk 2]
```diff
@@ -15,6 +15,18 @@ Disconnect volume node network of statefulset ${statefulset_id} for ${duration}
     ${node_name} =    get_volume_node    ${volume_name}
     disconnect_node_network    ${node_name}    ${duration}
 
+Disconnect volume nodes network for ${duration} seconds
+    [Arguments]    @{args}
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/keywords/network.resource:19]: Yes, we can figure out a standard way to document the parameters and usage of each keyword.

File: e2e/tests/cluster_restart.robot [hunk 3]
```diff
@@ -20,48 +22,66 @@ ${RETRY_COUNT}    300
...
     Given Set setting rwx-volume-fast-failover to ${RWX_VOLUME_FAST_FAILOVER}
-    And Create storageclass strict-local with    numberOfReplicas=1    dataLocality=strict-local
-    And Create storageclass nfs-4-2 with    nfsOptions=vers=4.2,noresvport,timeo=450,retrans=8
-    And Create persistentvolumeclaim 0 using RWO volume
-    And Create persistentvolumeclaim 1 using RWX volume
+    And Create storageclass longhorn-test with    dataEngine=${DATA_ENGINE}
+    And Create storageclass strict-local with    numberOfReplicas=1    dataLocality=strict-local    dataEngine=${DATA_ENGINE}
+    And Create storageclass nfs-4-2 with    nfsOptions=vers=4.2,noresvport,timeo=450,retrans=8    dataEngine=${DATA_ENGINE}
+    And Create storageclass nfs-hard-mount with    nfsOptions=hard,timeo=50,retrans=1    dataEngine=${DATA_ENGINE}
+    And Create storageclass nfs-soft-mount with    nfsOptions=soft,timeo=250,retrans=5    dataEngine=${DATA_ENGINE}
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/tests/cluster_restart.robot:35]: I'm not sure about this. The build-in storage class don't have `sc` prefix as well:

```
$ kubectl get sc
NAME                   PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
local-path (default)   rancher.io/local-path   Delete          WaitForFirstConsumer   false                  16m
longhorn (default)     driver.longhorn.io      Delete          Immediate              true                   11m
longhorn-static        driver.longhorn.io      Delete          Immediate              true                   11m
```

File: e2e/tests/cluster_restart.robot [hunk 4]
```diff
@@ -20,48 +22,66 @@ ${RETRY_COUNT}    300
 ${RETRY_INTERVAL}    1
 ${CONTROL_PLANE_NODE_NETWORK_LATENCY_IN_MS}    0
 ${RWX_VOLUME_FAST_FAILOVER}    false
+${DATA_ENGINE}    v1
 
 
 *** Test Cases ***
 Restart Cluster While Workload Heavy Writing
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/tests/cluster_restart.robot:29]: Sure! I'll add descriptions for each test case in the next PR. Thank you!

File: e2e/libs/keywords/network_keywords.py [hunk 5]
```diff
@@ -17,8 +19,23 @@ def cleanup_control_plane_network_latency(self):
         logging(f"Cleaning up control plane network latency")
         cleanup_control_plane_network_latency()
 
-    def disconnect_node_network(self, node_name, disconnection_time_in_sec):
-        disconnect_node_network(node_name, int(disconnection_time_in_sec))
+    async def disconnect_node_network(self, node_name, disconnection_time_in_sec):
+        await disconnect_node_network(node_name, int(disconnection_time_in_sec))
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/libs/keywords/network_keywords.py]: Good catch! The `await` has been removed. Thank you!

**Comment by yangchiu** [comment]
> nit: In general, if we can add brief description of test cases that will be helpful as the test cases are becoming bigger. 

Created https://github.com/longhorn/longhorn/issues/9204 for it.

File: e2e/libs/volume/crd.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (khushboo-rancher) [reply e2e/libs/volume/crd.py:423]: If checksum doesn't match, will the test case still pass or person implementing the test case will have to take care of it?
Btw, the message in the assert won't help?

[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/libs/volume/crd.py]: Good catch! I added the assertion back after the sleep. Thank you!

**Review by khushboo-rancher — APPROVED** [review]

**Pull request merged**